public/app/features/alerting/unified/rule-list/PaginatedDataSourceLoader.tsx

```diff
@@ -0,0 +1,128 @@
+import { groupBy } from 'lodash';
+import { useEffect, useMemo, useRef } from 'react';
+
+import { Dropdown, Icon, IconButton, Menu, Stack, Text } from '@grafana/ui';
+import { DataSourceRuleGroupIdentifier, ExternalRulesSourceIdentifier, RuleGroup } from 'app/types/unified-alerting';
+
+import { hashRule } from '../utils/rule-id';
+
+import { AlertRuleLoader } from './AlertRuleLoader';
+import { DataSourceSection, DataSourceSectionProps } from './components/DataSourceSection';
+import { LazyPagination } from './components/LazyPagination';
+import { ListGroup } from './components/ListGroup';
+import { ListSection } from './components/ListSection';
+import { usePrometheusGroupsGenerator } from './hooks/prometheusGroupsGenerator';
+import { usePaginatedPrometheusGroups } from './hooks/usePaginatedPrometheusGroups';
+
+export const GROUP_PAGE_SIZE = 40;
+
+interface PaginatedDataSourceLoaderProps extends Required<Pick<DataSourceSectionProps, 'application'>> {
+  rulesSourceIdentifier: ExternalRulesSourceIdentifier;
+}
+export function PaginatedDataSourceLoader({ rulesSourceIdentifier, application }: PaginatedDataSourceLoaderProps) {
+  const { uid, name } = rulesSourceIdentifier;
+  const prometheusGroupsGenerator = usePrometheusGroupsGenerator();
+
+  const groupsGenerator = useRef(prometheusGroupsGenerator(rulesSourceIdentifier, GROUP_PAGE_SIZE));
+
+  useEffect(() => {
+    const currentGenerator = groupsGenerator.current;
+    return () => {
+      currentGenerator.return();
+    };
+  }, [groupsGenerator]);
+
+  const {
+    page: groupsPage,
+    nextPage,
+    previousPage,
+    canMoveForward,
+    canMoveBackward,
+    isLoading,
+  } = usePaginatedPrometheusGroups(groupsGenerator.current, GROUP_PAGE_SIZE);
```

Comment:
I tried to create a more generalized approach, but it's challenging because of the variations in types—`prometheusGroupsGenerator` produces a generator of `PromRuleGroupDTO`, while `grafanaGroupsGenerator` yields a generator of `GrafanaPromRuleGroupDTO`. 
Moreover, `prometheusGroupsGenerator` requires the `rulesSourceIdentifier` parameter, which is not needed by grafana. 
I'm receptive to any ideas and additional modifications to the code. 
I didn't want to invest too much time on further generalization since I believe there comes a point where we should consider them as distinct entities.